lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/AvroPageDataReader.java

```diff
@@ -964,59 +869,59 @@ protected void decodeIntoBlock(Decoder decoder, BlockBuilder builder)
         }
     }
 
-    private static LongIoFunction<Decoder> getLongPromotionFunction(Schema writerSchema)
-            throws AvroTypeException
+    public static LongIoFunction<Decoder> getLongDecoderFunction(Schema writerSchema)
     {
-        if (writerSchema.getType() == Schema.Type.INT) {
-            return Decoder::readInt;
-        }
-        throw new AvroTypeException("Cannot promote type %s to long".formatted(writerSchema.getType()));
+        return switch (writerSchema.getType()) {
+            case INT -> Decoder::readInt;
+            case LONG -> Decoder::readLong;
+            default -> throw new IllegalArgumentException("Cannot promote type %s to long".formatted(writerSchema.getType()));
+        };
     }
 
-    private static FloatIoFunction<Decoder> getFloatPromotionFunction(Schema writerSchema)
-            throws AvroTypeException
+    public static FloatIoFunction<Decoder> getFloatDecoderFunction(Schema writerSchema)
     {
         return switch (writerSchema.getType()) {
             case INT -> Decoder::readInt;
             case LONG -> Decoder::readLong;
-            default -> throw new AvroTypeException("Cannot promote type %s to float".formatted(writerSchema.getType()));
+            case FLOAT -> Decoder::readFloat;
+            default -> throw new IllegalArgumentException("Cannot promote type %s to float".formatted(writerSchema.getType()));
         };
     }
 
-    private static DoubleIoFunction<Decoder> getDoublePromotionFunction(Schema writerSchema)
-            throws AvroTypeException
+    public static DoubleIoFunction<Decoder> getDoubleDecoderFunction(Schema writerSchema)
     {
         return switch (writerSchema.getType()) {
             case INT -> Decoder::readInt;
             case LONG -> Decoder::readLong;
             case FLOAT -> Decoder::readFloat;
-            default -> throw new AvroTypeException("Cannot promote type %s to double".formatted(writerSchema.getType()));
+            case DOUBLE -> Decoder::readDouble;
+            default -> throw new IllegalArgumentException("Cannot promote type %s to double".formatted(writerSchema.getType()));
         };
     }
 
     @FunctionalInterface
-    private interface LongIoFunction<A>
+    public interface LongIoFunction<A>
```

Comment:
Consider moving these to the classes like LongRead (can be follow up)

Reply:
Place to do a lot of shuffling when rewriting these pieces. Will happen in follow up